### ISSUE TITLE
AWS SSO Error after session expired

### ISSUE THREAD

**Issue by m-radzikowski** [Noovolari/leapp#108]
**Describe the bug**

In the morning I logged in through SSO. After the session expired, when I try to choose the account, I’m getting “AWS SSO Error”. Logs:

```
[2021-04-16 20:27:17.794] [info] [renderer] execute from Leapp:  {
  error: Error: Command failed: /usr/local/bin/az account clear 2>&1
      at ChildProcess.exithandler (child_process.js:312:12)
      at ChildProcess.emit (events.js:310:20)
      at maybeClose (internal/child_process.js:1021:16)
      at Socket.<anonymous> (internal/child_process.js:443:11)
      at Socket.emit (events.js:310:20)
      at Pipe.<anonymous> (net.js:672:12) {
    killed: false,
    code: 127,
    signal: null,
    cmd: '/usr/local/bin/az account clear 2>&1'
  },
  standardout: '/bin/sh: /usr/local/bin/az: No such file or directory\n',
  standarderror: ''
}
[2021-04-16 20:27:17.878] [error] [renderer] [t] UnauthorizedException: Session token not found or invalid UnauthorizedException: Session token not found or invalid
    at Object.extractError (file:///Applications/Leapp.app/Contents/Resources/app.asar/dist/leapp-client/main.a0a5735f897a9d1b1e5c.js:2:929950)
    at constructor.extractError (file:///Applications/Leapp.app/Contents/Resources/app.asar/dist/leapp-client/main.a0a5735f897a9d1b1e5c.js:2:875253)
    at constructor.callListeners (file:///Applications/Leapp.app/Contents/Resources/app.asar/dist/leapp-client/main.a0a5735f897a9d1b1e5c.js:2:2705757)
    at constructor.emit (file:///Applications/Leapp.app/Contents/Resources/app.asar/dist/leapp-client/main.a0a5735f897a9d1b1e5c.js:2:2705467)
    at constructor.emitEvent (file:///Applications/Leapp.app/Contents/Resources/app.asar/dist/leapp-client/main.a0a5735f897a9d1b1e5c.js:2:2889995)
    at constructor.e (file:///Applications/Leapp.app/Contents/Resources/app.asar/dist/leapp-client/main.a0a5735f897a9d1b1e5c.js:2:2885502)
    at r.runTo (file:///Applications/Leapp.app/Contents/Resources/app.asar/dist/leapp-client/main.a0a5735f897a9d1b1e5c.js:2:834476)
    at file:///Applications/Leapp.app/Contents/Resources/app.asar/dist/leapp-client/main.a0a5735f897a9d1b1e5c.js:2:834682
    at constructor.<anonymous> (file:///Applications/Leapp.app/Contents/Resources/app.asar/dist/leapp-client/main.a0a5735f897a9d1b1e5c.js:2:2885772)
    at constructor.<anonymous> (file:///Applications/Leapp.app/Contents/Resources/app.asar/dist/leapp-client/main.a0a5735f897a9d1b1e5c.js:2:2890051)
```

**To Reproduce**

1. Log in through SSO
2. Wait 12 hours
3. Try to change account under the same SSO (although I don't know if that's fully reproducible every time)

**Desktop (please complete the following information):**

 - OS: macOS
 - OS Version: 10.15.6
 - Leapp Version: 0.4.7

**Comment by andreacavagna01** [comment]
Will be fixed with the new AWS token generation with v0.5.0.

Closing the issue

**Comment by m-radzikowski** [comment]
@andreacavagna01 unfortunately, this still exists in v0.5.0. I logged in in the morning, now I'm getting exactly the same behavior and log.

How can I (manually) clear cache or whatever so I can select an account? Doing "log out" in Leapp does not help. And I really need to be able to use accounts.

**Comment by m-radzikowski** [comment]
Even when I restart Mac, it still does not work. I don't know why it works the next day 🤔

**Comment by sheldonhull** [comment]
I'm experiencing the same problem. I can login to the browser, but leapp is failing to allow me to regenerate my authentication through it.

**Comment by andreacavagna01** [comment]
The problem is the expiration time given by AWS SSO auth method that is not corrected, will work on this today, adding an extra limitation to re login again

**Comment by sheldonhull** [comment]
I haven't experienced the problem again so far. Might be able to close this now.

**Comment by andreacavagna01** [comment]
yes,  closing the issue

**Comment by m-radzikowski** [comment]
@andreacavagna01 I experienced this issue again, on v0.7.4.

Are logs still written to `~/Library/Logs/Leapp/log.log` on MacOS (https://docs.leapp.cloud/#logs)? Last log entries I have there are from 2021-06-30.

**Comment by ericvilla** [comment]
Hi @m-radzikowski, it is currently writing under //Library/Logs/Leapp/log.electronService.log, but that's definitely a bug. We'll fix it but, as far as now, please check the path that I've specified

**Comment by m-radzikowski** [comment]
My `~/Library/Logs/Leapp/log.electronService.log` contains only a single entry from 2021-11-17.

**Comment by m-radzikowski** [comment]
After the session expired, I cloned the repo and run the dev version of Leapp. I still couldn't connect, but now I have logs.

```
[2021-11-29 22:12:46.852] [info]  [renderer] [SessionCardComponent] Starting Session {
   "timestamp": "2021-11-29T21:12:46.852Z",
   "id": "46eae12e-f6af-4ca4-8eb7-cfbe050bfa4c",
   "account": "xxx",
   "type": "awsIamRoleChained"
}
[2021-11-29 22:12:47.094] [error] [renderer] Session token not found or invalid UnauthorizedException: Session token not found or invalid
    at Object.extractError (file:///Users/mradzikowski/Documents/leapp/dist/leapp-client/vendor.js:92867:27)
    at Request.extractError (file:///Users/mradzikowski/Documents/leapp/dist/leapp-client/vendor.js:93232:8)
    at Request.callListeners (file:///Users/mradzikowski/Documents/leapp/dist/leapp-client/vendor.js:95977:20)
    at Request.emit (file:///Users/mradzikowski/Documents/leapp/dist/leapp-client/vendor.js:95944:10)
    at Request.emit (file:///Users/mradzikowski/Documents/leapp/dist/leapp-client/vendor.js:94568:14)
    at Request.transition (file:///Users/mradzikowski/Documents/leapp/dist/leapp-client/vendor.js:93894:10)
    at AcceptorStateMachine.runTo (file:///Users/mradzikowski/Documents/leapp/dist/leapp-client/vendor.js:100217:12)
    at file:///Users/mradzikowski/Documents/leapp/dist/leapp-client/vendor.js:100229:10
    at Request.<anonymous> (file:///Users/mradzikowski/Documents/leapp/dist/leapp-client/vendor.js:93910:9)
    at Request.<anonymous> (file:///Users/mradzikowski/Documents/leapp/dist/leapp-client/vendor.js:94570:12)
```

I will be happy to provide more details, however, there is nothing more in the logs. But the behavior is reproducible - every day so far.

@ericvilla Can we reopen this issue?

**Comment by ericvilla** [comment]
For sure @m-radzikowski. I'm going to re-open it. Let's investigate the issue further and find a solution together

**Comment by PatrykMilewski** [comment]
I'm having the same issue after I tried using multiple SSO at the same time. All of the profiles from the new SSO are having this issue (old SSO works fine). I can't find logs in `~/Library/Logs` or `~/.aws` directories.

Error message from the app:
```
UnauthorizedException
Session token not found or invalid
```

Version: 0.8.0

**Comment by pethron** [comment]
Log files can be found here, we'll wait for some more info on this:

- [WINDOWS] - %USERPROFILE%\\AppData\\Roaming\\Leapp\\log.log
- [MACOS] - ~/Library/Logs/Leapp/log.log
- [LINUX] - ~/.config/Leapp/log.log

Link to docs: https://docs.leapp.cloud/troubleshooting/app-data/

**Comment by m-radzikowski** [comment]
On 0.8.0 on Mac the logs are missing. But I was able to get logs running the app from code - see above.

**Comment by urz9999** [comment]
Were your SSO sessions ported from a previous version of Leapp or were created directly from 0.8.0? There is still a possibility that the retro compatibility service was unable to correctly port your old configuration to the new one. Also for testing, try to rename leapp-lock.json to a backup, clean your keychain searching Leapp as a key and restart the new version of Leapp. Create the AWS SSO sessions as new ones and see what happens.

**Comment by m-radzikowski** [comment]
Those sessions were created in previous Leapp versions.

I removed the leapp-lock.json and records from keychain - thanks for the tip. I added sessions from scratch, so far I was able to connect. In worst case this is a backup plan 😉 

I will let you know tomorrow if I was able to reconnect after session expiration.

**Comment by m-radzikowski** [comment]
Ok, it seems to work now, with sessions added in v0.8.0. Is it some one-time bug or you think it may happen again with new Leapp versions?

I think we can close it.

**Comment by ericvilla** [comment]
In v0.8.0 Leapp checks whether the AWS SSO integration token - stored in the system vault - is expired or not, as you can see in the following line of code:

https://github.com/Noovolari/leapp/blob/5469354d11009f9348cb75aa864efb16899d82ae/src/app/services/aws-sso-integration.service.ts#L76

As Alex said "There is still a possibility that the retro compatibility service was unable to correctly port your old configuration to the new one"

We've introduced the concept of Leapp configuration version; that way, we'll be all able to implement atomic migrations from a configuration version to another, instead of understanding the current one using code conditions.

I would test this feature once again to be sure it is working correctly. May I ask you what kind of access methods were you running, so that we can effectively reproduce the bug

**Comment by m-radzikowski** [comment]
I tested it a few times over the last days, did not have any more problems.

**Comment by ericvilla** [comment]
Thank you for the feedback @m-radzikowski! I will leave this issue open till the end of this week.
If this issue does no more interest someone's installation, then I will close it

**Comment by m-radzikowski** [comment]
@ericvilla @urz9999 I updated to v0.8.1 yesterday and have problems again. Twice from yesterday.

First time after the update to v0.8.1. I removed the `~/.Leapp/Leapp-lock.json` file and set up sessions again from fresh. It worked but then (probably after 12 hours) I'm getting the error again.

I see two changes mentioned in the release notes related to AWS SSO, so they may be related:

- https://github.com/Noovolari/leapp/commit/369f6f8c15574ff7b26386cce24ebc9beca9c14e
- https://github.com/Noovolari/leapp/commit/fc09fcedc51141c52159bd82a7a47adda870a1f3

For now, I will revert to 0.8.0.

**Comment by ericvilla** [comment]
Hi @m-radzikowski, thank you for reporting! I'll check it out with @urz9999. Can you please confirm that the problem is still related to `UnauthorizedException: Session token not found or invalid`?
Thank you

**Comment by m-radzikowski** [comment]
Yes, exactly. Thanks.

**Comment by ericvilla** [comment]
@m-radzikowski I've tested the behaviour with an AWS SSO integration created from scratch. I forced it to expire after a minute, but it seems to work, i.e. it opens the browser for authentication. Even if I start a Chained Role from the expired AWS SSO one (AWS SSO access token expired), it opens the browser for authentication, as expected. Let me know if this is the behaviour that you expect to, or if I'm missing something. Thank you

**Comment by itsfeng** [comment]
I do see this behaviour as well. I'm using Version 8.0.1. Log files are not present under the mentioned location. Were able to resolve the issue by clicking "Log Out" in Leapp itself.

**Comment by ericvilla** [comment]
@m-radzikowski @itsfeng can you please check the presence of a key that starts with "aws-sso-integration-access-token-" in the keychain?

If this key is not present in the keychain - while pointed by Leapp configuration file - it generates an exception.

Thanks for the collaboration!

**Comment by m-radzikowski** [comment]
The issue is also reported by a few of my colleagues from the company, after upgrading to v0.8.1.

Today I got the same behavior on v0.8.0 🤷‍♂️ 

I decided to clean everything and update to v0.8.1. So far:

- removed `~/.Leapp/Leapp-lock.json` and `~/Library/Logs/Leapp`
- removed all Leapp keys from the MacOS keychain
- installed v0.8.1
- added AWS SSO integration and AWS IAM Role Chained session
- connection works
- in keychain new key `aws-sso-integration-access-token-xxx` is created
- `~/.Leapp/Leapp-lock.json` is created
- logs directory `~/Library/Logs/Leapp` is **not** created

I will update you at the end of the day, the issue occurs after token expiration after 12h.

> I forced it to expire after a minute, but it seems to work, i.e. it opens the browser for authentication.

@ericvilla how did you force it to expire?

**Comment by ericvilla** [comment]
Hi @m-radzikowski, thank you for the detailed response! The steps you followed to reset Leapp are right.
I've tested the behavior locally and it works; logs are written down into the log path, which is defined in https://github.com/Noovolari/leapp/blob/35c5f1c04e85ee7f167bea9a537f98884c4e30c8/src/app/services/app.service.ts#L78
I've logged the AWS SSO access token expiration time, which is persisted into the ./Leapp-lock.json file.
I logged in a few minutes ago, and the resulting log says:

> EXPIRATION TIME: Tue Feb 01 2022 19:25:01 GMT+0100 (Central European Standard Time)

In my previous test - not this one - I've forced the AWS SSO Integration's expiration by hard-coding an expired Date's ISOString in the payload of the following method invocation:
https://github.com/Noovolari/leapp/blob/35c5f1c04e85ee7f167bea9a537f98884c4e30c8/src/app/services/aws-sso-integration.service.ts#L89

I'm still investigating the reason behind the behaviour you reported. I'll keep you updated and provide you - and anyone who is encountering the same problem - a solution. Every suggestion is welcome!

**Comment by m-radzikowski** [comment]
It's 9 hours later. Token expired. Leapp did not open a browser window for SSO. No session is active.

When I click on the session, it does not connect, I get the UnauthorizedException.

The `~/Library/Logs/Leapp` still does not exist. So no logs I could share. @itsfeng also noted that there are no log files in his case, so I'm not the only one.

In MacOS keychain the same key still exists, the modification timestamp is from the morning.

There is nothing more I can provide you at the moment...

**Comment by ericvilla** [comment]
Hi @m-radzikowski and @itsfeng , we probably found what is causing the unwanted behaviour. We'll patch the existing implementation and test it today

**Comment by ericvilla** [comment]
@m-radzikowski was your pc always active (no sleep mode) during the 9 hours you specified in the last response?

**Comment by m-radzikowski** [comment]
@ericvilla I think so. I was working for most of the time, it was maybe blocked for some time, but not put to sleep mode.

**Comment by ericvilla** [comment]
@m-radzikowski @itsfeng I finally managed to find the root cause of the issue... hard to find, simple to explain.
The AWS SSO access token - that Leapp generates upon login - is bound to the browser session. If you log out from the browser, the AWS SSO access token - stored in Leapp - is no more valid.
To reproduce this problem:

- log into AWS SSO through Leapp (using external browser authentication);
- try to force sync AWS SSO integration, just to notice that the token is still valid;
- from the browser, log out from the AWS console;
- try to force sync AWS SSO integration again... you will get the error.

**Comment by ericvilla** [comment]
PS: we've to manage this exception and start a new login process if the access token is no more valid

**Comment by ericvilla** [comment]
We solved this issue in version 0.10.0 of Leapp! You can download it [here](https://www.leapp.cloud/releases)

**Comment by unfor19** [comment]
@ericvilla On the same topic, I'm getting similar error

```
[2023-07-25 10:13:44.770] [info]  [renderer] Starting Session
[2023-07-25 10:13:49.822] [warn]  [renderer] SAML authentication timeout exceeded
```


<img width="560" alt="image" src="https://github.com/Noovolari/leapp/assets/15122452/2de83a8a-3a25-4607-910c-b42ee3f7975d">

I'm on macOS 2019 amd64 Ventura 13.4.1; I've tried using all versions of Leapp, from 0.13+ to 0.18.1. I also tried removing everything from `~/.Leapp` and from `Keychain Access` that is related to Leapp; Still no success.

Previously, the app worked flawlessly; now, I'm not sure why it keeps popping up that error. My colleagues suffer from the same issue.

Any ideas on how to fix it?
